src/coreclr/vm/clsload.cpp

```diff
@@ -2801,7 +2801,10 @@ TypeHandle ClassLoader::PublishType(const TypeKey *pTypeKey, TypeHandle typeHnd)
         {
             MethodDesc * pMD = it.GetMethodDesc();
             CONSISTENCY_CHECK(pMD != NULL && pMD->GetMethodTable() == pMT);
-            if (!pMD->IsUnboxingStub())
+            // For {Task-returning, Async} variants of the same definition
+            // we associate the methoddef with the Task-returning variant since it
+            // matches the methadata signature.
+            if (!pMD->IsUnboxingStub() && !pMD->IsAsyncVariantMethod())
             {
```

Comment:
This could be interesting. Both variants have the same memberDef, so it is 2->1 relationship. 
However when going from memberDef to methoddesc is 1->1 map, and starting with memberDef will produce the non-async methoddesc - the one that matches the metadata signature.  (but it may not be the one that owns the IL!)